packages/malloy/src/tags.ts

```diff
@@ -0,0 +1,177 @@
+  properties: MalloyTagProperties;
+}
+interface DocTag {
+  docString: string;
+}
+type MalloyTag = DocTag | PropertyTag;
+function isDocTag(a: MalloyTag | undefined): a is DocTag {
+  return (a as DocTag)?.docString !== undefined;
+}
+
+export interface MalloyTags extends PropertyTag {
+  docStrings: string[];
+}
+
+export interface Taggable {
+  getTags: () => Tags;
+}
+
+/**
+ * Collection of useful functions for handing tag data ...
+ */
+export class Tags {
+  private tags: Annotation | undefined;
+  constructor(annotation: Annotation | undefined) {
+    this.tags = annotation;
+  }
+
+  getMalloyTags(): MalloyTags {
+    const ret: MalloyTags = {docStrings: [], properties: {}};
+    for (const tagLine of tagList(this.tags)) {
+      const tag = parseTag(tagLine, ret.properties);
+      if (isDocTag(tag)) {
+        ret.docStrings.push(tag.docString);
+      }
+    }
+    return ret;
+  }
+
+  getTagList(): string[] {
+    return tagList(this.tags);
+  }
+}
+
+function tagList(tagNode: Annotation | undefined): string[] {
+  if (!tagNode) {
+    return [];
+  }
+  return [
+    ...(tagNode.inherits ? tagList(tagNode.inherits) : []),
+    ...(tagNode.blockNotes || []),
+    ...(tagNode.notes || []),
+  ];
+}
+
+/**
+ * Lines which start '#"' are doc strings
+ * Lines which start '# ' are property lines ... The property parser is
+ * pretty simple, the langauge is:
+ * Property can be
+ *   name       -- set the value to "true"
+ *   -name      -- delete a value from the property
+ *   name=value -- assign value to the name
+ * Name can be any sequence of non space characters, or a string
+ * A value is any sequence of non space characters, or a string
+ * A string is " enclosing and ending in " with \" allowed inside
+ */
+function parseTag(
+  src: string,
+  tagProp: MalloyTagProperties
+): MalloyTag | undefined {
+  if (src.startsWith('#" ')) {
+    return {docString: src.slice(3)};
+  }
+  if (!src.startsWith('# ')) {
+    return;
+  }
+  /*
+   * I went back and forth if the grammar for these annotations should be
+   * in the parser or the lexer. Eventually the fact that the lexer ate
+   * newlines meant I couldn't figure out how to do it in the parser.
+   *
+   * Seems wrong to be writing a parser though.
+   */
+  const tokens = tokenize(src.slice(2));
+  let tn = 0;
+  const lastToken = tokens.length - 1;
+  while (tn <= lastToken) {
+    let token = tokens[tn];
+    if (token === '=') {
+      return undefined;
+    }
+    if (token[0] === '"') {
+      token = token.slice(1, -1);
+    }
+    if (tn !== lastToken && tokens[tn + 1] === '=') {
+      if (tn + 2 <= lastToken) {
+        let value = tokens[tn + 2];
+        if (value !== '=') {
+          if (value[0] === '"') {
+            value = value.slice(1, -1);
+          }
+          tagProp[token] = value;
+          tn += 3;
+          continue;
+        }
+      }
+      return undefined;
+    }
+    if (token.startsWith('-')) {
+      delete tagProp[token.slice(1)];
```

Comment:
Didn't know we were doing this, but that's neat. Seems useful.

Reply:
I love that we'll be able to do:

```
# dashboard
query: foo is ...

# -dashboard
query: bar is -> foo ...
```

Reply:
Yeah, I tried to tell lloyd that the usage was "render=barchart" , which would refine properly, but he made sad noises and this is how you do that.

I actually think "render=barchart" might be better, but the hierarchy is

"render-style-name"
can be refined with "refine-render=render-style-name"

But I thought I would give - a spin first